ebr.go

```diff
@@ -0,0 +1,263 @@
+)
+
+func (entityRules *EntityRules) ApplyOn(entities ...interface{}) error {
+	if err := entityRules.Validate(); err != nil {
+		return err
+	}
+	for _, e := range entities {
+		if err := entityRules.applyOnOne(e); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (entityRules *EntityRules) Validate() error {
+	if err := (&entityRules.Actions).validate(); err != nil {
+		return err
+	}
+	if err := (&entityRules.Conditions).validate(); err != nil {
+		return err
+	}
+	if rr := entityRules.Rules; rr == nil {
+		fmt.Println("Warning entity does not have rules")
+		return nil
+	} else {
+		for _, r := range rr {
+			if r.Name == "" {
+				return fmt.Errorf("Rule missing name: %v", r)
+			}
+			if r.When == nil {
+				fmt.Println("Warning rule: %s does not have When", r.Name)
+			}
+			if r.If == "" {
+				fmt.Println("Warning rule: %s does not have If", r.Name)
+			}
+			if r.Do == nil {
+				fmt.Println("Warning rule: %s does not have Do", r.Name)
+			}
+		}
+	}
+	return nil
+}
+
+func (mm *MethodsExecution) validate() error {
+	for _, m := range *mm {
+		if m.Name == nil {
+			return fmt.Errorf("cannot use empty name: %v", m)
+		}
+		if m.Method == nil {
+			return fmt.Errorf("cannot use empty method: %v", m)
+		}
+	}
+	return nil
+}
+
+func (entityRules *EntityRules) applyOnOne(entity interface{}) error {
+	re := reflect.ValueOf(entity)
+	for _, rule := range entityRules.Rules {
+		isTrigger, err := entityRules.isRuleTriggered(re, rule.When)
+		if err != nil {
+			return fmt.Errorf("errors occured in triggers check: %w", err)
+		}
+		if isTrigger {
+			isConditions, err := entityRules.entityHasConditions(re, rule.If)
+			if err != nil {
+				return fmt.Errorf("errors occured in conditions check: %w", err)
+			}
+			if isConditions {
+				if err := entityRules.execute(re, rule.Do); err != nil {
+					return fmt.Errorf("errors occured in actions execution: %w", err)
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (entityRules *EntityRules) isRuleTriggered(entity reflect.Value, when []string) (isTrigger bool, err error) {
+	//Check if at least one trigger is valid using When attribute of the current rule
+	isTrigger = true
+	for _, name := range when {
+		if isTrigger, err = entityRules.executeMethodBool(entity, name); err != nil {
+			return false, fmt.Errorf("%w: %s", err, name)
+		}
+		if isTrigger {
+			break
+		}
+	}
+	return isTrigger, nil
+}
+
+func (entityRules *EntityRules) entityHasConditions(entity reflect.Value, on string) (isConditions bool, err error) {
+	//Check if conditions are valid using If attribute of the current rule
+	if on == "" { //If no conditions are requiered
+		return true, nil
+	}
+	//Formating r.on for latter evaluation expression
+	listParser := []string{"not", "or", "and", "(", ")"}
+	namedConditions := make(map[string]bool)
+	var expression string
+	for _, name := range strings.Split(on, " ") {
```

Comment:
I think better to use https://play.golang.org/p/EcXszNg3TgG since it gonna allow to put` ( MyConditon` and `(MyCondition`

+ it removes Exists func need